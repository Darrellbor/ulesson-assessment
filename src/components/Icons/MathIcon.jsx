//@flow
import React, { type Node } from "react";

const MathIcon = (): Node => {
  return (
    <svg width="35" height="29" viewBox="0 0 35 29" fill="none">
      <g opacity="0.6">
        <path
          d="M34.3551 2.38304H21.2747L4.08657 26.9453L1.38144 16.8535"
          stroke="#FCFBF9"
          strokeWidth="1.9125"
          strokeMiterlimit="10"
          strokeLinejoin="round"
        />
        <path
          d="M19.3478 11.3513L31.7603 27.9945"
          stroke="#FCFBF9"
          strokeWidth="1.9125"
          strokeMiterlimit="10"
          strokeLinejoin="round"
        />
        <path
          d="M31.7603 11.3513L19.3478 27.9945"
          stroke="#FCFBF9"
          strokeWidth="1.9125"
          strokeMiterlimit="10"
          strokeLinejoin="round"
        />
        <path
          d="M1.16916 1.44641L5.33261 7.02724"
          stroke="#FCFBF9"
          strokeWidth="1.9125"
          strokeMiterlimit="10"
          strokeLinejoin="round"
        />
        <path
          d="M9.58513 1.44641L1.16916 12.7294"
          stroke="#FCFBF9"
          strokeWidth="1.9125"
          strokeMiterlimit="10"
          strokeLinejoin="round"
        />
      </g>
    </svg>
  );
};

export default MathIcon;
