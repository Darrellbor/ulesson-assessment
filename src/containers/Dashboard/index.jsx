import React, { Component } from "react";
import { Link } from "react-router-dom";

//context
import Context from "../../store/context";

//components
import Header from "../../components/Header";
import Button from "../../components/Button";

//icons
import MathIcon from "../../components/Icons/MathIcon";
import PhyIcon from "../../components/Icons/PhyIcon";
import ChemIcon from "../../components/Icons/ChemIcon";
import BioIcon from "../../components/Icons/BioIcon";
import EngIcon from "../../components/Icons/EngIcon";

class Dashboard extends Component {
  state = {
    lessons: [],
  };

  async componentDidMount() {
    await this.setSubjects();
    this.getLessons();
  }

  setSubjects = async () => {
    if (!this.context.subjects || this.context.subjects.length === 0) {
      await this.context.setSubjects();
    }
  };

  setIcon = (name) => {
    switch (name) {
      case "Mathematics":
        return <MathIcon />;
      case "Physics":
        return <PhyIcon />;
      case "Chemistry":
        return <ChemIcon />;
      case "Biology":
        return <BioIcon />;
      case "English":
        return <EngIcon />;
      default:
        break;
    }
  };

  setClassName = (name) => {
    switch (name) {
      case "Mathematics":
        return "math";
      case "Physics":
        return "phy";
      case "Chemistry":
        return "chem";
      case "Biology":
        return "bio";
      case "English":
        return "eng";
      default:
        break;
    }
  };

  getLessons = () => {
    if (
      this.context &&
      this.context.subjects &&
      this.context.subjects.data &&
      this.context.subjects.data.subjects
    ) {
      let lessons = [],
        temp = [],
        selected = [];
      const subjects = this.context.subjects.data.subjects;

      for (let subject of subjects) {
        for (let chapter of subject.chapters) {
          temp = [...temp, ...chapter.lessons];
        }
      }

      //mock recently watched topics
      for (let i = 0; i < 5; i++) {
        const randId = Math.floor(Math.random() * temp.length);

        if (selected.indexOf(randId) < 0) {
          lessons.push(temp[randId]);
        }

        selected.push(randId);
      }

      this.setState({ lessons });
    }
  };

  getSubject = (id) => {
    if (
      this.context &&
      this.context.subjects &&
      this.context.subjects.data &&
      this.context.subjects.data.subjects
    ) {
      const subjects = this.context.subjects.data.subjects;

      return subjects.filter((subject) => subject.id === parseInt(id, 10))[0];
    }
  };

  render() {
    const subjects =
      this.context &&
      this.context.subjects &&
      this.context.subjects.data &&
      this.context.subjects.data.subjects
        ? this.context.subjects.data.subjects
        : null;
    const lessons = this.state.lessons;

    return (
      <div className="Dashboard">
        <Header />
        <br />

        <div className="Dashboard__content">
          <h1>Hello Darrel,</h1>
          <br />
          <div className="Dashboard__subjects">
            {subjects &&
              subjects.map((subject) => (
                <Link key={subject.id} to={`/chapters/${subject.id}`}>
                  <div
                    className={`Dashboard__subject__${this.setClassName(
                      subject.name
                    )}`}
                  >
                    {this.setIcon(subject.name)}
                    <br />
                    <div
                      className={`Dashboard__subject__text ${
                        subject.name === "Mathematics" &&
                        "Dashboard__subject__text--math"
                      } `}
                    >
                      {subject.name === "English"
                        ? "English Language"
                        : subject.name}
                    </div>
                  </div>
                </Link>
              ))}
          </div>{" "}
          <br />
          <div className="Dashboard__recent">
            <div className="Dashboard__recent__header">
              <h3>Recently watched topics</h3>
              <Button type="button">See All</Button>
            </div>

            <div className="Dashboard__recent__items">
              {lessons &&
                lessons.map((lesson) => (
                  <Link
                    key={lesson.id}
                    to={`/video/${lesson.subject_id}/${lesson.chapter_id}/${lesson.id}`}
                  >
                    <div className="Dashboard__recent__item">
                      <div className="Dashboard__recent__item__img">
                        <img src={lesson.icon} alt="Comparison" />
                      </div>
                      <br />
                      <div
                        className={`Dashboard__recent__item__subject Dashboard__recent__item__subject--${this.setClassName(
                          this.getSubject(lesson.subject_id).name
                        )}`}
                      >
                        {this.getSubject(lesson.subject_id).name}
                      </div>
                      <div className="Dashboard__recent__item__title">
                        {lesson.name}
                      </div>
                    </div>
                  </Link>
                ))}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Dashboard.contextType = Context;

export default Dashboard;
