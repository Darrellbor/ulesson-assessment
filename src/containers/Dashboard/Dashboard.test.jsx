import React from "react";
import { shallow } from "enzyme";

import Dashboard from "./";

//Icons
import MathIcon from "../../components/Icons/MathIcon";
import PhyIcon from "../../components/Icons/PhyIcon";

describe("Dashboard Component", () => {
  let wrapper, instance;

  beforeEach(() => {
    wrapper = shallow(<Dashboard />);
    instance = wrapper.instance();
  });

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("Set Icon: Should return the appropriate icon based on subject name", () => {
    expect(instance.setIcon("Mathematics")).toEqual(<MathIcon />);
    expect(instance.setIcon("Physics")).toEqual(<PhyIcon />);
  });

  it("Set Class Name: Should return the appropriate short name based on subject name", () => {
    expect(instance.setClassName("Mathematics")).toBe("math");
    expect(instance.setClassName("Chemistry")).toBe("chem");
  });

  it("Get Lessons: Should set the lesson state value when called", () => {
    instance.getLessons();
    expect(wrapper.state("lessons").length).toBeGreaterThanOrEqual(0);
  });
});
