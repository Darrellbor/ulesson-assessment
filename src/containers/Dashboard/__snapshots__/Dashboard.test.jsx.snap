// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component renders correctly 1`] = `
<div
  className="Dashboard"
>
  <Header />
  <br />
  <div
    className="Dashboard__content"
  >
    <h1>
      Hello Darrel,
    </h1>
    <br />
    <div
      className="Dashboard__subjects"
    />
     
    <br />
    <div
      className="Dashboard__recent"
    >
      <div
        className="Dashboard__recent__header"
      >
        <h3>
          Recently watched topics
        </h3>
        <Button
          color="brand"
          type="button"
        >
          See All
        </Button>
      </div>
      <div
        className="Dashboard__recent__items"
      />
    </div>
  </div>
</div>
`;
